django_filters/fields.py

```diff
@@ -257,16 +269,21 @@ def __init__(self, *args, **kwargs):
 
         super().__init__(*args, **kwargs)
 
-    def _get_choices(self):
-        return super()._get_choices()
-
-    def _set_choices(self, value):
-        super()._set_choices(value)
-        value = self.iterator(self, self._choices)
-
-        self._choices = self.widget.choices = value
-
-    choices = property(_get_choices, _set_choices)
+    @property
+    def choices(self):
+        return super().choices
+
+    @choices.setter
+    def choices(self, value):
+        if DJANGO_50:
+            value = self.iterator(self, value)
+        else:
+            super()._set_choices(value)
+            value = self.iterator(self, self._choices)
+
+        # Simple `super()` syntax for calling a parent property setter is
+        # unsupported. See https://github.com/python/cpython/issues/59170
+        super(ChoiceIteratorMixin, self.__class__).choices.__set__(self, value)
```

Comment:
This is the most unfortunate bit, but works well nonetheless.

Ideally this whole override ought to be as simple as `super().choices = self.iterator(self, value)`.

But currently there is no support for calling the setter of a parent property, see https://github.com/python/cpython/issues/59170, although there does seem to be indication that a fix for this would be acceptable, see https://github.com/python/cpython/pull/29950#issuecomment-1122940064.

There is [this workaround](https://github.com/python/cpython/issues/59170#issuecomment-1093581240) which is even [used in Python itself](https://github.com/python/cpython/issues/59170#issuecomment-1093581243) but there were [challenges observed regarding mixins](https://github.com/python/cpython/issues/59170#issuecomment-1093581255). I have managed to make this work for this mixin by using `super(ChoiceIteratorMixin, self.__class__)` instead of `super(self.__class__, self.__class__)` which ensures that when calling e.g. `ChoiceField().choices =` (for a subclass of the mixin), the `super()` call will not be passed `ChoiceField` as the value for `self.__class__` as the first argument. Passing `ChoiceIteratorMixin` means that the next lookup in the MRO will be for `forms.ChoiceField.choices`.

Reply:
(I should also mention that this only works if we're using `@choices.setter` here and the property is defined here, not in the parent otherwise we end up with infinite recursion. Another reason why the getter must be redefined here.)